tests/test_security.py

```diff
@@ -134,9 +134,77 @@ def test_api_delete_authed(self):
         response = self.app.delete(url, headers=_auth("user2", "user2"))
         self.assertEqual(response.status_int, 200)
 
-    def test_api_rebuild_admin(self):
-        """/api/rebuild requires admin"""
+
+class TestEndpointSecurityAdmin(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        cls.package = package = make_package()
+        settings = {
+            "pyramid.debug_authorization": True,
+            "pypi.db": "tests.test_security.GlobalDummyCache",
+            "pypi.storage": "tests.test_security.GlobalDummyStorage",
+            "pypi.allow_delete": False,
+            "session.validate_key": "a",
+            "user.user": sha256_crypt.encrypt("user"),
+            "user.admin": sha256_crypt.encrypt("admin"),
+            "package.%s.user.user" % package.name: "rw",
+            "auth.admins": ["admin"],
+        }
+        app = main({}, **settings)
+        cls.app = webtest.TestApp(app)
+
+    def setUp(self):
+        cache = GlobalDummyCache()
+        cache.upload(
+            self.package.filename,
+            BytesIO(b"test1234"),
+            self.package.name,
+            self.package.version,
+        )
+
+    def tearDown(self):
+        GlobalDummyCache.global_packages.clear()
+        GlobalDummyStorage.global_packages.clear()
+        self.app.reset()
+
+    def test_api_pkg_delete_weak_user(self):
+        """/api/package/<pkg>/<filename> validates allow_delete setting"""
+
+        response = self.app.delete(
+            "/api/package/{package_name}/{filename}".format(
+                package_name=self.package.name, filename=self.package.filename
+            ),
+            expect_errors=True,
+            headers=_auth("user", "user"),
+        )
+        self.assertEqual(response.status_int, 403)
+
+    def test_api_pkg_delete_admin(self):
+        """/api/package/<pkg>/<filename> validates allow_delete setting"""
+        response = self.app.delete(
+            "/api/package/{package_name}/{filename}".format(
+                package_name=self.package.name, filename=self.package.filename
+            ),
+            headers=_auth("admin", "admin"),
+        )
+        self.assertEqual(response.status_int, 200)
+
+    def test_rebuild_weak_user(self):
+        """/admin/rebuild requires admin"""
+        response = self.app.get(
+            "/admin/rebuild/", expect_errors=True, headers=_auth("user", "user")
+        )
+        self.assertEqual(response.status_int, 403)
+
+    def test_rebuild_admin(self):
+        """/admin/rebuild requires admin"""
+        response = self.app.get("/admin/rebuild/", headers=_auth("admin", "admin"))
+        self.assertEqual(response.status_int, 200)
+
+    def test_api_unknown_url(self):
```

Comment:
Added this test since `/api/rebuild/` is a non-existant url, and I fixed the above tests to use `/admin/rebuild`, but it dropped the coverage of some logic. Adding this test restores the coverage to the expected level.